### PR TITLE
chore(xtask): Add `cargo xtask unpublished`

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This tries to make it easy to see what existing versions have not been published.  A future version of this could post to a PR what the current delta in version numbers for touched crates so reviewer have more context when deciding if they should ask for a crate version to be bumped
```console
$ cargo unpublished
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
     Running `/home/epage/src/personal/cargo/target/debug/xtask unpublished`
    Updating crates.io index
name published current
==== ========= =======
cargo-test-macro - 0.1.0
cargo-test-support - 0.1.0
cargo-util 0.2.3 0.2.4
mdman 0.0.0 0.1.0
resolver-tests - 0.1.0
cargo 0.70.0 0.71.0
cargo-credential 0.1.0 0.2.0
cargo-credential-1password 0.1.0 0.2.0
cargo-credential-gnome-secret 0.1.0 0.2.0
cargo-credential-macos-keychain 0.1.0 0.2.0
cargo-credential-wincred 0.1.0 0.2.0
benchsuite - 0.1.0
```

Room for improvement
- Aligning the start of each column
- Filtering the list by a commit range
- Adding this to an action to post to a review
- Maybe sorting the output
- Marking some our crates as `package.publish = false`, like benchsuite and resolver-tests

### How should we test and review this PR?

This is broken down commit by commit for easier seeing of the building blocks for our first xtask

### CODE REVIEW

File: .cargo/config.toml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+[alias]
+xtask = "run --package xtask --"
```
Comment (weihanglo) [inline .cargo/config.toml:2]: I thought you leaned toward multiple binaries (https://github.com/matklad/cargo-xtask/issues/2)

**Review by weihanglo — COMMENTED** [review]
It's both promising and unfortunate. I mean it's nice to have this by just reusing aargo-the-library. On the other hand we need to pull the whole library, increasing compile time a lot.

Anyhow, thank you for making this prototype so fast!

**Comment by epage** [comment]
> It's both promising and unfortunate. I mean it's nice to have this by just reusing aargo-the-library. On the other hand we need to pull the whole library, increasing compile time a lot.

I had considered writing this with `cargo metadata` but then I felt this would be a good oppotunity to dogfood.

**Comment by epage** [comment]
(in particular, using crates-index to walk the index only supports git and didn't want to enshrine that in our workflow)

File: .cargo/config.toml (hunk 1 above)
Comment (weihanglo) [reply .cargo/config.toml:2]: I wonder if we could build each binary on demand. `xtask` binary acts as a thin layer of parent command delegates other subcommands.

That is, when running `cargo xtask unpublished` it spawns a process running `cargo run -p xtask --bin unpublished`. With that we can probably avoid polling too many dependencies.

Comment (weihanglo) [reply .cargo/config.toml:2]: I am thinking of making `src/doc/build.sh` an xtask, which shouldn't compile the whole Cargo when running.

Comment (weihanglo) [reply .cargo/config.toml:2]: I was wrong. That mean we need `unpublished` to be a separate bin. However, I feel like we should go multi-binaries way to make build time less suck.

Comment (epage) [reply .cargo/config.toml]: I've gone ahead and created a dedicated binary

File: crates/xtask-unpublished/src/xtask.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
+use cargo::core::registry::PackageRegistry;
+use cargo::core::QueryKind;
+use cargo::core::Registry;
+use cargo::core::SourceId;
+use cargo::util::command_prelude::*;
+
+pub fn cli() -> clap::Command {
+    clap::Command::new("xtask-unpublished")
```
Comment (weihanglo) [inline crates/xtask-unpublished/src/xtask.rs:8]: We should document it. Let's do it after figuring out a better look of xtask for cargo.

**Review by weihanglo — APPROVED** [review]
This looks good to me. I don't have any particular view of what it would evolve over time at this moment. Keep in mind that xtask discoverability is not good enough in multi-binaries layout.

Given it is not really hard to tweak later, I'll merge as-is.

Thank you for kicking off this!

**Comment by weihanglo** [comment]
@bors r+

Note that the output format is not really suitable for CI job. Whoever going to get that done will need to make some changes.

**Pull request merged**